### PR TITLE
fix: support oci image index

### LINKED ISSUE #131: acr purge --untagged deletes manifests belonging to tagged multi-arch or signed images
**Describe the bug**

acr purge with --untagged flag delete manifests belonging to tagged images that are manifest lists (multi-arch image or image index), for example multi-arch images or signed images.

This broke ALL our images in Azure Container Registry.

**To Reproduce**

1. Prepare a simple Dockerfile:

    ```sh
    $ cat Dockerfile
    FROM docker.io/alpine:3
    ```

2. Build it as a multi-arch image and push it to our Azure Container Registry in an empty repository named `example/multi`:

    ```sh
    $ docker buildx build --platform linux/amd64,linux/arm64 -t myregistry.azurecr.io/example/multi:latest . --push
    [+] Building 4.3s (9/9) FINISHED
     => [internal] load .dockerignore                                                                                                      0.0s
     => => transferring context: 2B                                                                                                        0.0s
     => [internal] load build definition from Dockerfile                                                                                   0.0s
     => => transferring dockerfile: 61B                                                                                                    0.0s
     => [linux/amd64 internal] load metadata for docker.io/library/alpine:3                                                                0.9s
     => [linux/arm64 internal] load metadata for docker.io/library/alpine:3                                                                0.9s
     => CACHED [linux/amd64 1/1] FROM docker.io/library/alpine:3@sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a   0.0s
     => => resolve docker.io/library/alpine:3@sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a                      0.0s
     => CACHED [linux/arm64 1/1] FROM docker.io/library/alpine:3@sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a   0.1s
     => => resolve docker.io/library/alpine:3@sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a                      0.0s
     => exporting to image                                                                                                                 3.2s
     => => exporting layers                                                                                                                0.0s
     => => exporting manifest sha256:241d2510c991c5847944bac2dbc71f5a3a42760d66caae8b370af0cec830e546                                      0.0s
     => => exporting config sha256:b01c271483b8995dd0e56b367c47e6067880131a1efdfdc60d5561f206bcc666                                        0.0s
     => => exporting attestation manifest sha256:ac6816b2ab560a3c3e47a29b44c4435a4c312620da07f2f6d0589a8e4a097168                          0.0s
     => => exporting manifest sha256:42ca8fe0f16dfe78a006a1f9c2fb0ec17bc8f25b560d1d38b16f65f95f804754                                      0.0s
     => => exporting config sha256:9b49c0db52485ff8b9dfb5837d36f173e34d5320ee1b7dc912f7bfc5c231156a                                        0.0s
     => => exporting attestation manifest sha256:d74460be0584c8dfd87ef18a88ca8fdd5e35a62eacdc6a8051e85282c32fc7b3                          0.0s
     => => exporting manifest list sha256:e0516637706c08d2a9a7f2a34d52b70782b79480a3855f1f56c9917cedf32941                                 0.0s
     => => pushing layers                                                                                                                  1.4s
     => => pushing manifest for myregistry.azurecr.io/example/multi:latest@sha256:e0516637706c08d2a9a7f2a34d52b70782b79480a3855f1f56c99  1.7s
     => [auth] example/multi:pull,push token for myregistry.azurecr.io                                                                   0.0s
     => [auth] example/multi:pull,push tools/toolkit:pull token for myregistry.azurecr.io                                                0.0s
    ```

3. Check that we can pull it and then remove it from our local image cache so we can test what happens after running the `acr purge` command:

    ```sh
    $ docker pull myregistry.azurecr.io/example/multi:latest
    latest: Pulling from example/multi
    8921db27df28: Already exists
    Digest: sha256:e0516637706c08d2a9a7f2a34d52b70782b79480a3855f1f56c9917cedf32941
    Status: Downloaded newer image for myregistry.azurecr.io/example/multi:latest
    myregistry.azurecr.io/example/multi:latest

    $ docker image rm myregistry.azurecr.io/example/multi:latest
    Untagged: myregistry.azurecr.io/example/multi:latest
    Untagged: myregistry.azurecr.io/example/multi@sha256:e0516637706c08d2a9a7f2a34d52b70782b79480a3855f1f56c9917cedf32941
    Deleted: sha256:b01c271483b8995dd0e56b367c47e6067880131a1efdfdc60d5561f206bcc666
    ```

4. Run `acr purge` asking to delete tags older than 30 days and removing untagged manifests:

    ```sh
    $ az acr run --cmd "acr purge --filter 'example/multi:.*' --ago 30d --untagged" --timeout 3600 --registry myregistry /dev/null
    Queued a run with ID: cb1r
    Waiting for an agent...
    2023/02/03 11:28:46 Alias support enabled for version >= 1.1.0, please see https://aka.ms/acr/tasks/task-aliases for more information.
    2023/02/03 11:28:46 Creating Docker network: acb_default_network, driver: 'bridge'
    2023/02/03 11:28:46 Successfully set up Docker network: acb_default_network
    2023/02/03 11:28:46 Setting up Docker configuration...
    2023/02/03 11:28:47 Successfully set up Docker configuration
    2023/02/03 11:28:47 Logging in to registry: myregistry.azurecr.io
    2023/02/03 11:28:48 Successfully logged into myregistry.azurecr.io
    2023/02/03 11:28:48 Executing step ID: acb_step_0. Timeout(sec): 3600, Working directory: '', Network: 'acb_default_network'
    2023/02/03 11:28:48 Launching container with name: acb_step_0
    Deleting tags for repository: example/multi
    Deleting manifests for repository: example/multi
    Deleted myregistry.azurecr.io/example/multi@sha256:241d2510c991c5847944bac2dbc71f5a3a42760d66caae8b370af0cec830e546
    Deleted myregistry.azurecr.io/example/multi@sha256:42ca8fe0f16dfe78a006a1f9c2fb0ec17bc8f25b560d1d38b16f65f95f804754
    Deleted myregistry.azurecr.io/example/multi@sha256:d74460be0584c8dfd87ef18a88ca8fdd5e35a62eacdc6a8051e85282c32fc7b3
    Deleted myregistry.azurecr.io/example/multi@sha256:ac6816b2ab560a3c3e47a29b44c4435a4c312620da07f2f6d0589a8e4a097168

    Number of deleted tags: 0
    Number of deleted manifests: 4
    2023/02/03 11:28:51 Successfully executed container: acb_step_0
    2023/02/03 11:28:51 Step ID: acb_step_0 marked as successful (elapsed time in seconds: 3.014514)

    Run ID: cb1r was successful after 6s
    ```

    At this point you can observe it didn't delete tags, however it removed 4 manifests which doesn't make sense as we've just uploaded the multi-arch image and the repo `example/multi` was empty.

5. Try to pull the image now:

    ```sh
    $ docker pull myregistry.azurecr.io/example/multi:latest
    latest: Pulling from example/multi
    manifest for myregistry.azurecr.io/example/multi:latest not found: manifest unknown: manifest sha256:241d2510c991c5847944bac2dbc71f5a3a42760d66caae8b370af0cec830e546 is not found
    ```

    The tag exist but references to a manifest that has been deleted...

**Expected behavior**

`acr purge` shouldn't remove manifests belonging to tagged multi-arch images.

**Screenshots**

N/A

**Any relevant environment information**

 - OS: N/A
 - Version N/A

**Additional context**

Signed non multi-arch images with `cosign` reproduce the same behavior as signed images use the same manifest list format for referencing the image and the signature. In fact we've discovered this issue working with images signed with `cosign`.

I'm guessing that `acr purge` is unable to determine that a manifest is in use by a tag when it is indirectly referenced from a tag that is a manifest list.

Here I attach some samples of the manifests for multi-arch, signed and "normal or plain" images:

<details><summary>Manifest for a multi-arch image</summary>
<p>

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 480,
      "digest": "sha256:241d2510c991c5847944bac2dbc71f5a3a42760d66caae8b370af0cec830e546",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 480,
      "digest": "sha256:42ca8fe0f16dfe78a006a1f9c2fb0ec17bc8f25b560d1d38b16f65f95f804754",
      "platform": {
        "architecture": "arm64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 566,
      "digest": "sha256:ac6816b2ab560a3c3e47a29b44c4435a4c312620da07f2f6d0589a8e4a097168",
      "annotations": {
        "vnd.docker.reference.digest": "sha256:241d2510c991c5847944bac2dbc71f5a3a42760d66caae8b370af0cec830e546",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 566,
      "digest": "sha256:d74460be0584c8dfd87ef18a88ca8fdd5e35a62eacdc6a8051e85282c32fc7b3",
      "annotations": {
        "vnd.docker.reference.digest": "sha256:42ca8fe0f16dfe78a006a1f9c2fb0ec17bc8f25b560d1d38b16f65f95f804754",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      }
    }
  ]
}
```

</p>
</details>

<details><summary>Manifest for a signed image with `cosign`</summary>
<p>

```json

{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 2943,
      "digest": "sha256:2d583b83ee23e2a6daeb3470c9e35d89714c4a5e63fe838b922134b57ae0cd54",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 566,
      "digest": "sha256:da4ef47c5d8c0d579f3dd101a9405085b32859f5f0c15372762962ef102ca521",
      "annotations": {
        "vnd.docker.reference.digest": "sha256:2d583b83ee23e2a6daeb3470c9e35d89714c4a5e63fe838b922134b57ae0cd54",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      }
    }
  ]
}
```

</p>
</details>

<details><summary>Manifest for a normal/plain image</summary>
<p>

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType": "application/vnd.docker.container.image.v1+json",
    "size": 1472,
    "digest": "sha256:042a816809aac8d0f7d7cacac7965782ee2ecac3f21bcf9f24b1de1a7387b769"
  },
  "layers": [
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 3370628,
      "digest": "sha256:8921db27df2831fa6eaa85321205a2470c669b855f3ec95d5a3c2b46de0442c9"
    }
  ]
}
```

</p>
</details>

### PR DESCRIPTION
**Pull request by northtyphoon**
**Purpose of the PR**

- add complete oci and docker image media-type to the accept header

Fixes #131

NOTE: #130 added the multi-arch support for oci image index but the command could still run into the error `[{"errors":[{"code":"MANIFEST_UNKNOWN","message":"OCI index found, but accept header does not support OCI indexes"}]}]`. The pr addresses the problem.

### CODE REVIEW

**Comment by sajayantony** [comment]
Overall no concerns on this PR. Just wanted to confirm if we have a case where a manifest in the index has a tag, that we don't delete that manifest.

**Review by sajayantony — APPROVED** [review]

**Comment by northtyphoon** [comment]
@sajayantony thanks for asking, if the manifest has any tag, it won't be deleted unless the tag itself is also been purged.

File: cmd/api/acrsdk.go [hunk 1]
```diff
@@ -21,10 +21,21 @@ import (
...
-	registryURL           = ".azurecr.io"
-	manifestTagFetchCount = 100
-	manifestV2ContentType = "application/vnd.docker.distribution.manifest.v2+json"
+	prefixHTTPS                      = "https://"
+	registryURL                      = ".azurecr.io"
+	manifestTagFetchCount            = 100
+	manifestORASArtifactContentType  = "application/vnd.cncf.oras.artifact.manifest.v1+json"
+	manifestOCIArtifactContentType   = "application/vnd.oci.artifact.manifest.v1+json"
+	manifestOCIImageContentType      = "application/vnd.oci.image.manifest.v1+json"
+	manifestOCIImageIndexContentType = "application/vnd.oci.image.index.v1+json"
```
Comment (shizhMSFT) [inline cmd/api/acrsdk.go:30]: Should we reuse the media types defined in https://github.com/opencontainers/image-spec/blob/main/specs-go/v1/mediatype.go

Comment (shizhMSFT) [reply cmd/api/acrsdk.go:30]: We did not do that for docker media types because we don't want to introduce the dependency on distribution.

Comment (northtyphoon) [reply cmd/api/acrsdk.go]: updated

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**